genienlp/run_generation.py

```diff
@@ -379,101 +577,130 @@ def run_generation(args):
+                o = o[batch_input_sequence_lengths[i % batch_size]:]
+                # print('original context tokens: ', batch_context_tokens[i % batch_size])
+                # print('original input sequence: ', batch_input_sequences[i % batch_size])
+
+                if args.stop_tokens is not None:
+                    min_index = len(o)
+                    for stop_token_id in stop_token_ids:
+                        try:
+                            index = o.index(stop_token_id)
+                            min_index = min(index, min_index)
+                        except ValueError:
+                            pass
+                    if min_index < len(o) and o[min_index] == tokenizer.convert_tokens_to_ids('?'):
+                        # always include the question mark
+                        min_index = min_index + 1
+                    if min_index < len(o) and o[min_index] == tokenizer.convert_tokens_to_ids(args.stop_tokens[0]):
+                        # include </paraphrase> in logit calculation
+                        o_logits = o_logits[:len(o_logits)-(len(o)-min_index-1)]
+                    o = o[:min_index]
+                
+                text = tokenizer.decode(o, clean_up_tokenization_spaces=True, skip_special_tokens=False)
+
+                # assert tokenizer.pad_token not in text
+                text = text.replace(tokenizer.pad_token, '')
+                text = re.sub('\s\s+', ' ', text) # remove duplicate white spaces
+                text = text.strip()
+                if not args.skip_heuristics:
+                    text = output_heuristics(text, batch_reverse_maps[i % batch_size])
+                batch_outputs[i % batch_size].append(text)
+
+                if args.selection_criterion == 'bleu':
+                    # computeBLEU always converts to lower case first, so do not worry about lower/upper case here
+                    criterion = computeBLEU([text], [[batch_input_sequences[i % batch_size]]])
+                else:
+                    criterion = np.mean(o_logits)
+                batch_criterion[i % batch_size].append(criterion)
+                # print('generated tokens: ', o)
+                # print('o_logits = ', o_logits)
+                # print('generated cirterion: %.2f' % criterion)
+                # print('text = ', text)
+                # print('-'*10)
+
+
+        if args.selection_criterion == 'none':
+            all_outputs.extend(batch_outputs)
+        else:
+            for idx, example in enumerate(batch_outputs):
+                print('input sequence: ', batch_input_sequences[idx % batch_size])
+                c, example = tuple(zip(*sorted(list(zip(batch_criterion[idx], example)), reverse=True)))
+                print(example)
+                print(c)
+                print('-'*10)
+                selection = example[0]
+                all_outputs.append([selection])
```

Comment:
This whole function needs to be split into multiple functions, it's already way too long and too nested to read.